src/api/lib/backend/api/build_results/binaries.rb

```diff
@@ -64,16 +64,14 @@ def self.wipe_published_locked(project_name)
         # the name of the binary as keys and the architectures as the value
         def self.transform_binary_packages_response(response)
           list = {}
+          list.default = []
           parsed_response = Xmlhash.parse(response)
           return list if parsed_response.blank?
           packages = [parsed_response["packages"]].flatten
           packages.each do |build|
             architectures = [build["arch"]].flatten
             packages = [build["name"]].flatten
-            packages.each do |package|
-              architectures = architectures.concat(list[package]) if list[package]
-              list[package] = architectures.uniq
-            end
+            packages.each { |package| list[package] |= architectures }
```

Comment:
Why not leaving the ``concat``? I find this ``|=`` harder to understand than a concat!

```
packages.each { |package| list[package].concat(architectures.to_a).uniq }
```

If it is because of the nil check you could do a ``to_a`` before. Also the ``uniq`` is more explicit. I understand that the ``|=`` is smart but I find it harder to understand ...

Reply:
Why is it difficult to understand?? :scream: 

It is just the [Set Union for arrays in Ruby](https://ruby-doc.org/core-2.2.0/Array.html#method-i-7C), which is exactly what we want to do here. If that had been written like this in the first place, the bug wouldn't have been introduced. Doing `concat` + `uniq` when that is exactly the set union. 

And it is not because of the nil check, as with the default `[]` it is not needed anymore. And the `to_a` you are suggesting is also not needed. This will work as well:

``` Ruby
packages.each { |package| list[package].concat(architectures).uniq }

Reply:
It is harder to understand than the union as this is nothing what I do every day. While first looking at the code I thought you do ``||=`` so I thought this doesn't make sense and tried it in IRB to figure out that this performs a Union.
When looking at  ``` list[package].concat(architectures).uniq``` on the other hand it is immediately clear that this concatenates the arrays and makes them uniq.
The ``to_a`` seems indeed not necessary as we wrap the architecture always in ``[]``.flatten.

I totally admit that the union is clever but I find it quite obfuscating! And as performance shouldn't be a problem here, I would prefer the explicit way of writing this 😄

Reply:
And the bug was not caused because of the Union or concat.uniq (as this is the same as you already wrote). It was caused because we overwrote the variable set outside of the loop.

Reply:
If you have so strong feelings about it I will approve but I still thing ``concat.uniq`` is more explicit and more readable. I would prefer if you write at least
```
packages.each { |package| list[package] = list[package] | architectures }
```
to avoid mixing it up with ``||=``

Reply:
@ChrisBr 

> It is harder to understand than the union as this is nothing what I do every day. While first looking at the code I thought you do ``||=`` so I thought this doesn't make sense and tried it in IRB to figure out that this performs a Union.

`|` is in all (or most) programming languages the bitwise operator... wich in an array can only be the union. 

> to avoid mixing it up with `||=`

If `|=` can be mixing up with `||=`, then `|` and `||` can also be mixing up... :confused: 



It was clear for me, but if it is not clear for you I will leave it as it was before, I do not want to obfuscate the code for anybody. I will think if there is any other way to refactor this... :thinking:

Reply:
> If |= can be mixing up with ||=, then | and || can also be mixing up...

Nope! The point is: We use ~500 times ``||=`` compared to once ``|=``.

Reply:
also @ChrisBr, to write it as you want this should be `uniq!` because `uniq` doesn't modify the object, so:

``` Ruby
packages.each { |package| list[package].concat(architectures.to_a).uniq! }
```

But that this doesn't work as well. An example:

``` Ruby
[45] pry(main)> list = Hash.new([])
=> {}
[46] pry(main)> list[:a] = [1,2,3]
=> [1, 2, 3]
[47] pry(main)> architectures = [2,3,4]
=> [2, 3, 4]
[49] pry(main)> packages = [:a,:b]
```
``` Ruby
[52] pry(main)> packages.each { |package| list[package].concat(architectures).uniq! }
```

And everything seems to work:
``` Ruby
[56] pry(main)> list[:a]
=> [1, 2, 3, 4]
[57] pry(main)> list[:b]
=> [2, 3, 4]
```
But:
``` Ruby
[54] pry(main)> list
=> {:a=>[1, 2, 3, 4]}
[58] pry(main)> list[:c]
=> [2, 3, 4]
```

What we have done is modifying the default and not changing the value of b :rofl: Not sure if this is on purpose of a bug :thinking: 

So it can not be written as you want, so I will leave it as before. Using `concat` and `uniq` there is not other way.

Reply:
Discussed offline!